Faster cyclic

Sentient07 opened the pull request:
rebase of gh-40

- [ ]  Make a Theano flag fast_inplace that default to False. Use the new new algo only when it is True.
- [ ]  Make sure all tests pass when the flag is False
- [ ]  Make sure most tests pass when the flag is True. For failing test, investigate it this is a normal failure with the new algo. Is so, change the test to expect the failure when using the new algo and expect the old bahavior when using the old algo.
- [ ] Make an even faster version that only check the changed part of the graph by optimization, not the full graph.

nouiz commented on theano/gof/destroyhandler.py:
```diff
@@ -790,7 +846,9 @@ def on_import(self, fgraph, app, reason):
 
         # If it's a destructive op, add it to our watch list
         if getattr(app.op, 'destroy_map', {}):
+            # TODO: check here only one level of fast destroy_map.
```
remove that comment. It is now done.

Sentient07 replied:
@nouiz I'm sorry, am just seeing this. I didn't get notification from github

nouiz commented on theano/gof/destroyhandler.py:
```diff
@@ -777,6 +779,60 @@ def on_detach(self, fgraph):
+                continue
+            # Current simplest version don't allow destroyed inputs to
+            # have more then 1 client.
+            raise InconsistencyError()
```
Change this line to:

self.fail_validate = True

nouiz commented on theano/gof/destroyhandler.py:
```diff
@@ -807,7 +813,8 @@ def fast_destroy(self, app):
             # Current simplest version don't allow destroyed inputs to
             # have more then 1 client.
-            raise InconsistencyError()
+                self.fail_validate = True
```
bad indentation.

nouiz replied:
I'll make a PR to your PR. Don't change this.

Sentient07 commented on theano/gof/tests/test_destroyhandler.py:
```diff
@@ -241,43 +292,91 @@ def test_destroyers_loop():
+        except InconsistencyError:
+            return
+        else:
+            inconsistent(g)
```
Okay, did I make mistake here ? The removed code shows  `consistent(g)`  and I am checking for inconsistency ?

Sentient07 replied:
@nouiz yeah, that seems to work. i don't know how and why I even did a check for inconsistency. I changed that back and the tests are all fine. I will push it and update the PR 👍

Sentient07 commented on theano/gof/tests/test_destroyhandler.py:
```diff
@@ -241,43 +292,91 @@ def test_destroyers_loop():
+        except InconsistencyError:
+            return
+        else:
+            inconsistent(g)
```
@nouiz If this line seems to cause assertion error in the recent built, that means line 372 didn't raise InconsistencyError which used to (until the previous commit) on the newer algorithm. And as said in previous comment, it should have been `consistent(g)` instead of `inconsistent(g)` ..

Sentient07 commented:
@nouiz there were some additional commits than what i had added recently. So i rebased it, but by mistake with master. So there are huge number of commits. I'm sorry. 
Can I open a PR to original repo itself ?
